recipes-bsp/acpi-tables/samples/edison/leds.asl

```diff
@@ -0,0 +1,88 @@
+/*
+ * Intel Edison
+ *
+ * This table is modified from the Minnowboard Turbot leds.asl for Edison.
+ * 
+ * According to Intel® Edison Kit for Arduino Hardware Guide March 2017
+ * DS2 is the standard LED on the Arduino* board.
+ * It runs using the ‘blink’ code or whenever Digital I/O 13 is asserted High. 
+ * It can be used as an indicator under direct control.
+ * 
+ * This adds GPIO LEDs device for the DS2 LED and sets default heartbeat
+ * trigger for that.
+ *
+ * In Linux you need to set CONFIG_LEDS_GPIO=y (or m) to be able to use
+ * this device.
+ *
+ * Copyright (C) 2016, Intel Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+DefinitionBlock ("leds.aml", "SSDT", 5, "", "LEDS", 1)
```

Comment:
... The LED is connected to GP40, but not really. It's connected to DIG13_1P8. And this is connected to either GP40 or GP109 depending on SPI_CLK_SEL (U16 IO1.3). So when we enable SPI SPI_CLK_SEL (ssp5-clk-mux) is hogged and LED won't work.

Can we make the function selectable at run time other than not hogging SPI_CLK_SEL?

Reply:
@andy-shev what to do here? Leave it for now?

Reply:
If you decide we can leave it for now, I have tested leds.aml:
 
<details>
  <summary>Test results with/without spi enabled & with/without leds.aml loaded</summary>
**** This with ACPI_FEATURES_edison ?= "uart_2w i2c spi":

```
root@edison:~# blink-led 
Traceback (most recent call last):
  File "/usr/bin/blink-led", line 82, in <module>
    set_line("SPI_CLK_SEL", 0)
  File "/usr/bin/blink-led", line 66, in set_line
    line.request(consumer=line.owner().name(), type=gpiod.LINE_REQ_DIR_OUT)
OSError: [Errno 16] Device or resource busy
```

-> SPI_CLK_SEL is hogged by arduino.asl

With ACPI_FEATURES_edison ?= "uart_2w i2c"

```
root@edison:~# blink-led 
Traceback (most recent call last):
  File "/usr/bin/blink-led", line 86, in <module>
    led.request(consumer=led.owner().name(), type=gpiod.LINE_REQ_DIR_OUT)
OSError: [Errno 16] Device or resource busy
root@edison:~# modprobe -i ledtrig-heartbeat
```

-> GPIO40 hogged by leds.asl

```
modprobe -i ledtrig-heartbeat
echo 1 > /sys/class/leds/heartbeat/invert
```
-> LED blinks with heartbeat
```
rmdir /sys/kernel/config/acpi/table/leds 
```

-> does not release GPIO40

**** Booting without loading leds.aml

`blinkled 
`
blinks led by toggling GPIO40 from user space

</details>

Reply:
IIRC *leds.aml* is compatible with `gpio-leds` driver which serves platform device. To unload it properly we need my hack-patch. It seems it's not being shared.
Regarding to the run-time changing function — it's not possible for hogged pins.
I just realized that we might solve TRI_STATE_ALL issue whenever we will have support for ACPI r6.1 in pin control library. In that case we might write an ASL which contains pin function and configuration information. But it's not a nearest future.

Reply:
So, for now I just need to document not to use the led in combination with spi - or reboot when switching function. Noted.